File: May-LeetCoding-Challenge/23-Interval-List-Intersections/Interval-List-Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] A, int[][] B) {
        int res[][];
        List<int[]> resList = new ArrayList();
        
        int i = 0, j = 0;
        int first = 0, second = 0;
        
        while (i < A.length && j < B.length) {
            first = Math.max(A[i][0], B[j][0]);
            second = Math.min(A[i][1], B[j][1]);
            if (first <= second) {
                resList.add(new int[]{first, second});
            }
            
            if (A[i][1] < B[j][1])
                i++;
            else 
                j++;
        }
        
        return resList.toArray(new int[resList.size()][]);
    }
}

File: May-LeetCoding-Challenge/23-Interval-List-Intersections/Interval-List-Intersections.scala
object Solution {
  def intervalIntersection(A: Array[Array[Int]], B: Array[Array[Int]]): Array[Array[Int]] = {
    import scala.collection.mutable
    var res = mutable.Seq.empty[Array[Int]]
    var indexA = 0
    var indexB = 0
    while (indexA < A.length && indexB < B.length) {
      val (intAStart, intAEnd) = (A(indexA)(0), A(indexA)(1))
      val (intBStart, intBEnd) = (B(indexB)(0), B(indexB)(1))
      val start = intAStart max intBStart
      val end = intAEnd min intBEnd
      if (intAEnd >= intBEnd) indexB += 1
      if (intBEnd >= intAEnd) indexA += 1
      if (end >= start) res = res :+ Array(start, end)
    }
    res.toArray
  }
}